File: scala_study/src/main/scala/leetcode/reorderList.scala
package leetcode

class ListNode(_x: Int, _next: ListNode) {
  var x = _x
  var next = _next
}

object Solution{
  def reorderList(head: ListNode): Unit= {
    var p = head
    var len: Int  = 0
    while (p != null){
      len += 1
      p = p.next
    }
    p=head
    val l1 = new ListNode(0,head)
    val l2 = new ListNode(0,null)

    for (_ <- 1 until ( len / 2 + len % 2)){
      p=p.next
    }
    //截取前半段列表
    val tail = p
    p=p.next
    tail.next = null

    // 反转后半段链表
    while (p != null) {
      //头插法逆置链表
      val pPost = p.next
      p.next = l2.next
      l2.next = p
      p= pPost
    }

    //合并两个链表

    var p1 = l1.next
    p = l2.next
    while (p != null && p1 != null){
      val p1Post = p1.next
      val pPost = p.next
      p1.next = p
      p.next = p1Post

      p1 = p1Post
      p = pPost
    }

  }

}

File: JavaLearn/src/main/java/com/leetcode/ReOrderList.java
package com.leetcode;

public class ReOrderList {
    /**
     *方法二：寻找链表中点 + 链表逆序 + 合并链表
     * 注意到目标链表即为将原链表的左半端和反转后的右半端合并后的结果。
     *
     * 这样我们的任务即可划分为三步：
     *
     * 找到原链表的中点（参考「876. 链表的中间结点」）。
     *
     * 我们可以使用快慢指针来 O(N)O(N) 地找到链表的中间节点。
     * 将原链表的右半端反转（参考「206. 反转链表」）。
     *
     * 我们可以使用迭代法实现链表的反转。
     * 将原链表的两端合并。
     *
     * 因为两链表长度相差不超过 11，因此直接合并即可。
     *
     * @param head
     */
    public void reorderList(ListNode head) {
        if (head == null){
            return;
        }

//      找到原链表的中点（参考「876. 链表的中间结点」）。
        ListNode slow=head,fast=head;
        while (fast.next != null && fast.next.next != null){
            slow =slow.next;
            fast = fast.next.next;
        }

//      将原链表的右半端反转（参考「206. 反转链表」）。
        ListNode mid = slow.next;
        slow.next = null;

        ListNode dump = new ListNode(0,null);
        ListNode p = mid;

        while (p != null){
            ListNode pPost = p.next;
            p.next = dump.next;
            dump.next = p;

            p = pPost;
        }

//      将原链表的两端合并。
        fast = head;
        p = dump.next;

        while (fast != null || p != null){

            ListNode fastTemp = fast.next;
            ListNode pTemp = p.next;

            fast.next = p;
            p.next =fastTemp;

            fast = fastTemp;
            p = pTemp;
        }

        return;
    }
}
